### PR TITLE
gs-9.53 transparency (fixes #862); blend mode; transparency groups 

### LINKED ISSUE #862: ghostscript transparency changes coming
Not exactly a bug, but anyway ...

As you may know, Ghostscript's transparency support has been in flux. GS developer Chris Liddell recently posted a summary of the proposed changes to come (https://ghostscript.com/pipermail/gs-devel/2020-May/010295.html), and Bruno Voisin posted a summary of the history (https://tug.org/pipermail/tex-live/2020-May/045752.html).

With the next gs release (9.53), I guess it would be good for pgf/tikz to support the new operators discussed in the mail, if they are present, since the old ones will disappear in a year or two.

Hope this helps somehow. --best, karl.

### PR DESCRIPTION
**Pull request by agrahn**
**1.** GS 9.52 introduced new transparency operators

````
.setfillconstantalpha
.setstrokeconstantalpha
````

Cf. issue #862.

They are supposed to replace `.setopacityalpha` and allow opacity to be set individually for stroking and filling operations. `.setopacityalpha` will be deprecated with the upcoming version 9.53. This PR makes use of the new operators, but keeps `.setopacityalpha` as a fallback.

**2.** This PR also implements Transparency Groups and Blend Modes.

----

The code listing below contains examples from the PGF manual. Typeset with:
````
latex example.tex
dvips example
ps2pdf -dALLOWPSTRANSPARENCY example.ps
````
Outputs from `ps2pdf` and `pdftex` for comparison:
[transparencyExamples-dvips.pdf](https://github.com/pgf-tikz/pgf/files/4846538/transparencyExamples-dvips.pdf)
[transparencyExamples-pdftex.pdf](https://github.com/pgf-tikz/pgf/files/4846539/transparencyExamples-pdftex.pdf)

(Poppler-based viewers don't render correctly the knockout example in the `pdftex` output, but MuPDF and AcroReader do.)

````latex
\documentclass{article}
\pagestyle{empty}

\usepackage{tikz}
\usetikzlibrary {patterns,shapes.symbols}

\begin{document}
  \begin{tikzpicture}[thick,fill opacity=0.5]
    \filldraw[fill=red] (0:1cm) circle (12mm);
    \filldraw[fill=green] (120:1cm) circle (12mm);
    \filldraw[fill=blue] (-120:1cm) circle (12mm);
  \end{tikzpicture}
%
  \begin{tikzpicture}[every node/.style={fill,draw}]
    \draw[line width=2mm,blue!50,line cap=round] (0,0) grid (3,2);
    \node[opacity=0.5] at (1.5,2) {Upper node};
    \node[draw opacity=0.8,fill opacity=0.2,text opacity=1] at (1.5,0) {Lower node};
  \end{tikzpicture}
%
  \begin{tikzpicture}[fill opacity=0.5]
    \fill[red] (0,0) circle (1);
    \fill[red] (1,0) circle (1);
  \end{tikzpicture}

  transparency group:
  \begin{tikzpicture}[opacity=0.5]
    \begin{scope}[transparency group]
      \draw [line width=5mm] (0,0) -- (2,2);
      \draw [line width=5mm] (2,0) -- (0,2);
    \end{scope}
  \end{tikzpicture}

  \begin{tikzpicture}
    \pattern[pattern=checkerboard,pattern color=black!15](-1,-1) rectangle (3,1);
    \node at (0,0) [forbidden sign,line width=2ex,draw=red,fill=white] {Smoking};
    \begin{scope}[transparency group,opacity=.5]
      \node at (2,0) [forbidden sign,line width=2ex,draw=red,fill=white] {Smoking};
    \end{scope}
  \end{tikzpicture}
%
  \begin{tikzpicture}
    \pattern[pattern=checkerboard,pattern color=black!15](-1,-1) rectangle (3,1);
    \node at (0,0) [forbidden sign,line width=2ex,draw=red,fill=white] {Smoking};
    \begin{scope}[transparency group,opacity=.5]
      \node (s) at (2,0) [forbidden sign,line width=2ex,draw=red,fill=white] {Smoking};
      \draw [opacity=.5, line width=2ex, blue] (1.2,0) -- (2.8,0);
    \end{scope}
  \end{tikzpicture}

  blend-mode `screen':
  \tikz {
    \begin{scope}[transparency group]
      \begin{scope}[blend mode=screen]
        \fill[red!90!black] ( 90:.6) circle (1);
        \fill[green!80!black] (210:.6) circle (1);
        \fill[blue!90!black] (330:.6) circle (1);
      \end{scope}
    \end{scope}
  }

  knockout example:
  \begin{tikzpicture}
    \shade [left color=red,right color=blue] (-2,-1) rectangle (2,1);
    \begin{scope}[transparency group=knockout]
      \fill [white] (-1.9,-.9) rectangle (1.9,.9);
      \node [opacity=0,font=\fontencoding{T1}\fontfamily{ptm}\fontsize{45}{45}\bfseries] {Ti\emph{k}Z};
    \end{scope}
  \end{tikzpicture}
  
\end{document}
````

### CODE REVIEW

File: tex/generic/pgf/systemlayer/pgfsys-dvips.def [hunk 1]
```diff
@@ -138,6 +150,28 @@
...
+%
+\def\pgfsys@transparencygroupfrombox#1{%
+  \special{ps:
+    mark /NamespacePush pdfmark % limit `{pgfxform}' name visibility
+    mark /_objdef {pgfxform} /BBox [clippath pathbbox newpath] /BP pdfmark
+    save gsave
+    /.setstrokeconstantalpha where
+      {pop 1 .pgfsetfillopacityalpha 1 .pgfsetstrokeopacityalpha} if
+  }%
+  \hskip\pgf@picminx\rlap{\smash{\box#1}}%
```
Comment (hmenke) [inline tex/generic/pgf/systemlayer/pgfsys-dvips.def]: Relying on the definition of `\rlap` and `\smash` is maybe not a good idea.  For example `amsmath` modifies the definition of `\smash`.
```suggestion
  \ht#1=0pt
  \dp#1=0pt
  \hskip\pgf@picminx\hbox to 0pt{\box#1\hss}%
```

**Review by hmenke — CHANGES_REQUESTED** [review]
Could you please also rebase onto master and fix your email address in the commit?  It doesn't resolve to your GitHub username.

File: tex/generic/pgf/systemlayer/pgfsys-dvips.def [hunk 2]
```diff
@@ -161,8 +170,13 @@
     /.setstrokeconstantalpha where
       {pop 1 .pgfsetfillopacityalpha 1 .pgfsetstrokeopacityalpha} if
   }%
+<<<<<<< HEAD
+  \ht#1=0pt%
+  \dp#1=0pt%
+=======
   \ht#1=0pt
   \dp#1=0pt
+>>>>>>> 972751ed77acd7187b40d7b4da1c9d617e7abda0
```
Comment (Mo-Gul) [inline tex/generic/pgf/systemlayer/pgfsys-dvips.def]: Here seems to be something left behind ...
```suggestion
  \ht#1=0pt
  \dp#1=0pt
```

**Review by Mo-Gul — CHANGES_REQUESTED** [review]
At least a minor change is required.

File: tex/generic/pgf/systemlayer/pgfsys-dvips.def (hunk 2 above)
Comment (agrahn) [reply tex/generic/pgf/systemlayer/pgfsys-dvips.def]: I fixed it, hopefully. Henry asked me to rebase this onto master. Is `git rebase master` the correct command?

**Comment by agrahn** [comment]
> Could you please also rebase onto master ...

Is `git rebase master` the necessary command for this (while being on `pstransparency` branch)?

**Comment by Mo-Gul** [comment]
Unfortunately I am not a git expert, but I am pretty sure that @ilayn knows the right commands.

**Comment by hmenke** [comment]
https://docs.github.com/en/github/using-git/about-git-rebase
https://git-scm.com/book/en/v2/Git-Branching-Rebasing

**Comment by hmenke** [comment]
Rebased and merged in 5a4bb8cfdc91e35c57a41fc287d7a85c38dfb3b6

**Pull request closed without merging**